rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricFileFacts.java

```diff
@@ -108,66 +134,228 @@ public void close(ISourceLocation loc) {
+            var messages = new ArrayList<Diagnostic>();
+            for (var versionedDiagnostics : list) {
+                if (versionedDiagnostics != null) {
+                    messages.addAll(versionedDiagnostics.messages);
+                }
+            }
+            return messages;
+        }
+    }
+
     private class FileFact {
         private final ISourceLocation file;
-        private volatile List<Diagnostic> parseMessages = Collections.emptyList();
-        private volatile List<Diagnostic> typeCheckerMessages = Collections.emptyList();
-        private final ParametricSummaryBridge summary;
+
+        // To replace old diagnostics when new diagnostics become available in a
+        // thread-safe way, we need to atomically: (1) check if the version of
+        // the new diagnostics is greater than the version of the old
+        // diagnostics; (2) if so, replace old with new. This is why
+        // `AtomicReference` and `VersionedDiagnostics` are needed here.
+        private final AtomicReference<VersionedDiagnostics> diagnosticsParser = new AtomicReference<>();
+        private final AtomicReference<VersionedDiagnostics> diagnosticsAnalyzer = new AtomicReference<>();
+        private final AtomicReference<VersionedDiagnostics> diagnosticsBuilder = new AtomicReference<>();
+
+        private final ParametricSummaryBridge summaryAnalyzer;
+        private final ParametricSummaryBridge summaryBuilder;
+
+        private final AtomicInteger latestVersionCalculateAnalyzer = new AtomicInteger();
+        private final AtomicInteger latestVersionCalculateBuilder = new AtomicInteger();
+
+        @SuppressWarnings("java:S3077")
+        private volatile @MonotonicNonNull CompletableFuture<Optional<VersionedDiagnostics>> calculateAnalyzerLatestResult;
 
         public FileFact(ISourceLocation file) {
             this.file = file;
-            this.summary = new ParametricSummaryBridge(exec, file, columns, contrib, lookupState);
+            this.summaryAnalyzer = new ParametricSummaryBridge(exec, file, columns, contrib, lookupState, contrib::analyze);
+            this.summaryBuilder = new ParametricSummaryBridge(exec, file, columns, contrib, lookupState, contrib::build);
         }
 
         public void reloadContributions() {
-            summary.reloadContributions();
+            summaryAnalyzer.reloadContributions();
+            summaryBuilder.reloadContributions();
         }
 
-        private void reportTypeCheckerMessages(List<Diagnostic> messages) {
-            typeCheckerMessages = messages;
-            sendDiagnostics();
+        private VersionedDiagnostics reportDiagnostics(AtomicReference<VersionedDiagnostics> box,
+                VersionedTree tree, List<Diagnostic> messages) {
+
+            var newer = new VersionedDiagnostics(tree, messages);
+            if (VersionedDiagnostics.setIfNewer(box, newer)) {
+                sendDiagnostics();
+            }
+            return newer;
```

Comment:
if not newer, should we not return the old one? aka, the one that won the race?

Reply:
This method has two responsibilities: (1) create a new `VersionedDiagnostics` object, which is needed by the caller regardless of whether it replaced the current diagnostics; (2) try to replace the current diagnostics with the new one (if newer) and send the diagnostics to the client.

I'm not extremely happy with putting these two responsibilities in one method, but I feel the alternatives (splitting in two methods or constructing the new object at the call site) aren't much better (but not much worse either). To better convey intent, I renamed the method and added some comments, but I'm fine to go with one of the alternatives too.